test/e2e_node/device_plugin_failures_pod_status_test.go

```diff
@@ -0,0 +1,256 @@
+			return err
+		}, f.Timeouts.PodDelete, f.Timeouts.Poll).Should(gomega.MatchError((gomega.ContainSubstring("not found"))))
+	})
+
+	ginkgo.It("will report a Device Status for the failed pod in the pod status", func(ctx context.Context) {
+		// randomizing so tests can run in parallel
+		resourceName := fmt.Sprintf("test.device/%s", f.UniqueName)
+		devices := []kubeletdevicepluginv1beta1.Device{{ID: "testdevice", Health: kubeletdevicepluginv1beta1.Healthy}}
+		plugin := testdeviceplugin.NewDevicePlugin(nil)
+
+		err := plugin.RegisterDevicePlugin(ctx, f.UniqueName, resourceName, devices)
+		defer plugin.Stop() // should stop even if registration failed
+		gomega.Expect(err).To(gomega.Succeed())
+
+		ginkgo.By("initial state: capacity and allocatable are set")
+		gomega.Eventually(getNodeResourceValues, nodeStatusUpdateTimeout, f.Timeouts.Poll).WithContext(ctx).WithArguments(resourceName).Should(gomega.Equal(ResourceValue{Allocatable: 1, Capacity: 1}))
+
+		// schedule a pod that requests the device
+		client := e2epod.NewPodClient(f)
+		pod := client.Create(ctx, createPodWrongImage(resourceName, 1))
+
+		// wait for the pod to be running
+		gomega.Expect(e2epod.WaitForPodCondition(ctx, f.ClientSet, pod.Namespace, pod.Name, "Back-off pulling image", f.Timeouts.PodStartShort,
+			func(pod *v1.Pod) (bool, error) {
+				if pod.Status.Phase == v1.PodPending &&
+					len(pod.Status.ContainerStatuses) > 0 &&
+					pod.Status.ContainerStatuses[0].State.Waiting != nil &&
+					pod.Status.ContainerStatuses[0].State.Waiting.Reason == "ImagePullBackOff" {
+					return true, nil
+				}
+				return false, nil
+			})).To(gomega.Succeed())
+
+		expectedStatus := []v1.ResourceStatus{
+			{
+				Name: v1.ResourceName(resourceName),
+				Resources: []v1.ResourceHealth{
+					{
+						ResourceID: "testdevice",
+						Health:     v1.ResourceHealthStatusHealthy,
+					},
+				},
+			},
+		}
+
+		gomega.Eventually(func() []v1.ResourceStatus {
+			pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			return pod.Status.ContainerStatuses[0].AllocatedResourcesStatus
+		}, devicePluginUpdateTimeout, f.Timeouts.Poll).Should(gomega.Equal(expectedStatus))
+
+		// now make the device unhealthy
+		devices[0].Health = kubeletdevicepluginv1beta1.Unhealthy
+		plugin.UpdateDevices(devices)
+
+		expectedStatus[0].Resources[0] = v1.ResourceHealth{
+			ResourceID: "testdevice",
+			Health:     "Unhealthy",
+		}
+
+		gomega.Eventually(func() []v1.ResourceStatus {
+			pod, err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			return pod.Status.ContainerStatuses[0].AllocatedResourcesStatus
+		}, devicePluginUpdateTimeout, f.Timeouts.Poll).Should(gomega.Equal(expectedStatus))
+
+		// deleting the pod
+		err = f.ClientSet.CoreV1().Pods(f.Namespace.Name).Delete(ctx, pod.Name, metav1.DeleteOptions{})
+		gomega.Expect(err).To(gomega.Succeed())
+
+		// wait for the pod to be deleted
```

Comment:
`waitForContainerRemoval` should provide stronger guarantees

Reply:
thank you! We may need to move this one to the test common, I looked for it =)